### PR TITLE
DBZ-3951 Adding a safety check, isValid, for connection failure (incremental snapshot)

### PR DESCRIPTION
**Pull request by vivekwassan**
**Context:** 
At-times facing `com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure" exception`

**Steps to reproduce:**
1. Start an incremental snapshot and let it finish
2. Try the same snapshot after sometime (~30 mins)
3. Getting "{{com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure" exception}}

**Solution**
Adding a safety check, isValid, to check if the connection object is valid.

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 1]
```diff
@@ -874,7 +875,7 @@ public synchronized boolean isConnected() throws SQLException {
         if (conn == null) {
             return false;
         }
-        return !conn.isClosed();
+        return !conn.isClosed() && conn.isValid(CONNECTION_VALID_CHECK_TIMEOUT_IN_SEC);
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java:878]: Hum, hum. This seems to issue a statement, and we call `isConnected()` a lot. I'm wondering whether we should rather have a separate method `JdbcConnection::isValid()` and call that in the snapshot change event source above? @jpechane, any thoughts?

**Comment by gunnarmorling** [comment]
Hey @vivekwassan, can you confirm whether the Postgres connector tests pass with this change locally for you?

**Comment by vwassan** [comment]
> Hey @vivekwassan, can you confirm whether the Postgres connector tests pass with this change locally for you?

@gunnarmorling On local, when I run 
```
mvn clean install -B -pl debezium-connector-postgres -am -Passembly,wal2json-decoder -Ddebezium.test.records.waittime=5 -Dformat.formatter.goal=validate -Dformat.imports.goal=check -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn -Dmaven.wagon.http.pool=false -Dmaven.wagon.httpconnectionManager.ttlSeconds=120
```
Test does not pass and it gets stuck as following and nothing happens

Screen Shot 2021-09-03 at 4 16 51 PM

**Comment by vivekwassan** [comment]
Hi @gunnarmorling 👋 , as Postgres connector tests are [failing](https://github.com/debezium/debezium/pull/2662#issuecomment-912788510) (unsure of the reason) and inspired from [your comment](https://github.com/debezium/debezium/pull/2662#discussion_r699975621), I added a separate isValid() in my [latest](https://github.com/debezium/debezium/pull/2662/commits/866daeba913015f1e30205b4ceee168bf1717024) commit. 
Calling this method only for `mysql` flow. 
Could you share your opinion?

cc @jpechane

**Comment by jpechane** [comment]
@vivekwassan Hi, would the same problem be present also for `SignalBasedIncrementalSnapshotChangeEventSource`?

**Comment by vivekwassan** [comment]
> @vivekwassan Hi, would the same problem be present also for `SignalBasedIncrementalSnapshotChangeEventSource`?

Hi @jpechane 👋 , Good point! My team, at Shopify, is using Mysql read-only incremental snapshot only, so we faced connection failure in the flow of `MySqlReadOnlyIncrementalSnapshotChangeEventSource` only. 
Now I feel, it might be occurring for `SignalBasedIncrementalSnapshotChangeEventSource` as it similarly uses `MySqlConnection`.
 So, [moved](https://github.com/debezium/debezium/pull/2662/commits/bc27c25bd16b8eeb3d32389be6965ea609bc1b94) check for `isValid` in `AbstractIncrementalSnapshotChangeEventSource`.

**Comment by jpechane** [comment]
@vivekwassan Applied, thanks.

**Pull request merged**